src/CodeGen_OpenCL_Dev.cpp

```diff
@@ -313,6 +461,35 @@ void CodeGen_OpenCL_Dev::CodeGen_OpenCL_C::visit(const Cast *op) {
     }
 }
 
+void CodeGen_OpenCL_Dev::CodeGen_OpenCL_C::visit(const Select *op) {
+    if (op->condition.type().is_vector()) {
+        // We eliminated boolean vectors in OpenCL C. In the process
+        // of doing so, we replaced the vector conditions of selects
+        // with integer vectors, compared with 0. To make legal OpenCL
+        // C, we just need to grab the integer vector out of the
+        // comparison with zero.
+        Expr condition = op->condition;
+        const NE* cond_neq = condition.as<NE>();
+        if (cond_neq) {
+            if (is_zero(cond_neq->b)) {
+                condition = cond_neq->a;
+            }
+        }
```

Comment:
Actually, the assert is even too aggressive to handle the selects generated by CodeGen_C. For example, absd generates a0 > a1 as the condition. This is fine because operator > in OpenCL C for vectors returns an integer vector. However, Halide thinks its a bool vector, which triggers the assert.